File: mpmt1.cpp
/*
 * mpmt1.cpp: A simple example of C++ threading and multiprocessing.
 *
 * License:
 *   Apache License, Version 2.0
 * History:
 *   2024/06/02 v0.1 Initial version
 * Author:
 *   Masanori Itoh <masanori.itoh@gmail.com>
 * BUILD:
 *   * `$ g++ -o mpmt1cpp mpmt1.cpp`
 * TODO:
 *   * Use getopt
 *   * Write multi process version.
 */
#include <iostream>
#include <thread>
#include <string>
#include <chrono>
#include <list>
#include <unistd.h>

using namespace std::chrono;

void busy_worker(uint64_t duration_ms) {
  
  uint64_t start_ts = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
  while (true) {
    uint64_t current_ts = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
    if (current_ts - start_ts >= duration_ms) {
      //Looks like "Expired." and endl below are separatedly send to stdout.
      //std::cout << "Expired."  << std::endl;
      //Atomic output preferred.
      printf("Expired.\n");
      return;
    }
  }
}


int main(int argc, char **argv){
  int num_context = 3;
  int duration = 5;
  std::list<std::thread>tlist;

  if (argc >= 2) {
    num_context = std::stoi(argv[1]);
  }
  if (argc >= 3) {
    duration = std::stoi(argv[2]);
  }
  //printf("num_context: %d duration: %d\n", num_context, duration);
  std::cout << "num_context: " << num_context << " duration: " << duration << std::endl;

  for (int i = 0; i < num_context; i++) {
    std::thread t = std::thread(busy_worker, duration * 1000);
    std::cout << "created thread...: " << t.get_id() << std::endl;
    tlist.push_back(std::move(t));
  }

  std::list<std::thread>::iterator it;
  for (it = tlist.begin(); it != tlist.end(); ++it){
    std::cout << "joining thread...: " << (*it).get_id() << std::endl;
    (*it).join();
  }
  return 0;
}

File: mpmt1.js
// mpmt1.js: A Node.js version of mpmt1.py
//
//License:
//  Apache License, Version 2.0
//History:
//  * 2021/12/30 v0.1 Initial version
//Author:
//  Masanori Itoh <masanori.itoh@gmail.com>
//TODO:
//  * Implement async mode(?)

function busy_worker(timeout) {

    // ts and ts_save are in mili-seconds
    var ts_save = Date.now()
    // duration is in seconds.
    var max = timeout * 1000

    while (true) {
        ts = Date.now()
        if ((ts - ts_save) >= max) {
            console.log('Expired! ' + (ts - ts_save))
            break
        }
    }
    if (mode == 'p') {
        process.exit(0)
    }
}


var posix_getopt = require('posix-getopt')
var parser, option

const cluster = require('cluster')

var num_context = 4
var duration = 10
var mode = 'p'

parser = new posix_getopt.BasicParser('n:d:m:', process.argv)
while ((option = parser.getopt()) !== undefined) {
    switch (option.option) {
    case 'n':
        num_context = option.optarg
        break;
    case 'd':
        duration = option.optarg
        break;
    case 'm':
        mode = option.optarg
        break;
    }
}


if (mode == 'p') {
    //Process mode (cluster)
    if (cluster.isMaster) {
        console.log('num_context: ' + num_context + ' duration: ' + duration + ' mode: ' + mode)
        console.log('parent: pid= ' + process.pid)

        for (var i = 0; i < num_context; i++) {
            cluster.fork()
        }

        cluster.on('exit', (code, signal) => {
            // signal is the value for process.exit()? node code?
            console.log('on exit called. code= ' + code + ' signal= ' + signal)
        });

    } else {
        console.log('child: pid=' + process.pid)
        busy_worker(duration)
    }
} else {
    //Thread mode (worker_thread)
    const {
        Worker,
        isMainThread,
        //parentPort,
    } = require("worker_threads");
    
    if (isMainThread) {
        console.log('num_context: ' + num_context + ' duration: ' + duration + ' mode: ' + mode)
        const threads = [];
        for (var i = 0; i < num_context; i++) {
            const worker = new Worker(__filename, {argv: process.argv.slice(2)});
            threads.push(worker)
        }
        //TODO(thatsdone): Implement join() equivalent for synchronization
    } else {
        busy_worker(duration)
    } 
}    
